src/jgikbase/idmapping/storage/mongo/id_mapping_mongo_storage.py

```diff
@@ -44,29 +45,79 @@ def __init__(self, db: Database) -> None:
         self._ensure_indexes()
 
     def _ensure_indexes(self):
-        self.db[_COL_USERS].create_index([(_FLD_AUTHSOURCE, 1), (_FLD_USER, 1)], unique=True)
+        self.db[_COL_USERS].create_index(_FLD_USER, unique=True)
         self.db[_COL_USERS].create_index(_FLD_TOKEN, unique=True)
 
-    def create_or_update_local_user(self, user: User, token: HashedToken) -> None:
+    def create_local_user(self, user: User, token: HashedToken) -> None:
```

Comment:
Almost all the code in this file is about error handling, and so little of it is about intentional functionality (two lines?). Are you going to have to do all this duplication checking for creating/updating anything in the db? It seems like a ton of overhead.

Reply:
I don't really have concrete changes to request on this one. I think the code here is not very readable, but I trust that you find all of this plumbing necessary in order to work with mongo, since I have little experience with mongo + python. To me, this code does not clearly desribe functionality related to any real features (are these users admins of namespaces?). It would be nice to see actual HTTP endpoints and to see what happens with real request data.

Maybe it's possible to handle all these duplicate key errors in a single Flask error handler.

Reply:
In my experience it's pretty typical that error handling code >> happy path code. There's only one way, usually, for the happy path to work, and lots of ways for things to go wrong. Hence, more code.

That's why all of this stuff is hidden behind a DAL so users of the DAL don't have to worry about it and get a clear interface (which isn't currently the case, but I'm building up to it).

> Are you going to have to do all this duplication checking for creating/updating anything in the db?

If it's an error to have a duplicate of X in the database, then yeah, there's going to have to be code to handle that case.

> To me, this code does not clearly desribe functionality related to any real features (are these users admins of namespaces?)

I don't really get that - the design doc describes the use case here: https://github.com/jgi-kbase/IDMappingService/blob/develop/design/design.md#design

> It would be nice to see actual HTTP endpoints and to see what happens with real request data.

This is a DAL implementation, and so is not necessarily running in a server context. These particular functions will not be run in the server given the current design (other than get_user(), obviously).

> Maybe it's possible to handle all these duplicate key errors in a single Flask error handler.

There's no guarantee this code is running in Flask, and coupling MongoDB and Flask wouldn't be a great design choice as there's talk about switching the DB to Cassandra or something else. That's why the DAL implementation translates mongo-specific errors to general errors, so that a Cassandra (say) DAL implementation can do the same thing, and we can swap out the DAL implementation without changing any other code.